packages/flame/lib/src/components/core/component.dart

```diff
@@ -806,21 +823,40 @@ class Component {
     onMount();
     _setMountedBit();
     _lifecycleManager?.finishMounting();
-    if (!existingChild) {
-      _parent!.children.add(this);
-    }
-    if (_children != null) {
-      _children!.forEach(
-        (child) => child._mount(parent: this, existingChild: true),
-      );
-    }
+    _parent!.children.add(this);
+    _reAddChildren();
     _lifecycleManager?.processQueues();
+    _parent!.onChildrenChanged(this, ChildrenChangeType.added);
+  }
+
+  /// Used by [_reAddChildren].
+  static final List<Component> _tmpChildren = [];
+
+  /// At the end of mounting, we remove all children components and then re-add
+  /// them one-by-one. The reason for this is that before the current component
+  /// was mounted, its [children] may have contained components in arbitrary
+  /// state -- initial, loading, unmounted, etc. However, we don't want to
+  /// have such components in a component tree. By removing and then re-adding
+  /// them, we ensure that they are placed in a queue, and will only be placed
+  /// into [children] once they are fully mounted.
+  void _reAddChildren() {
+    if (_children != null && _children!.isNotEmpty) {
+      assert(_tmpChildren.isEmpty);
```

Comment:
is this to avoid some race condition?

Reply:
This is a precondition: I expect `_tmpChildren` to be empty at this point (and if not, the logic would fail). It shouldn't be possible to have this list non-empty here, because I'm clearing it after every use, on line 851. Still, if someone were to add a new method that also uses `_tmpChildren`, leaving it non-empty, then this assert would alert of danger.